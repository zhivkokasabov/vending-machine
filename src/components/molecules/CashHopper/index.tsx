import React from "react";

import { Coin } from "components/atoms/Coin";
import { useCashHopper } from "providers/CashHopper.provider";

export const CashHopper = () => {
  const { stashedCoins } = useCashHopper();

  const stackDefaultValue: Record<string, number> = {};
  const groupedCoins = stashedCoins.reduce((acc, curr) => {
    const roundedKey = curr.toFixed(2);

    acc[roundedKey] = acc[roundedKey] ? acc[roundedKey] + 1 : 1;

    return acc;
  }, stackDefaultValue);
  // might sort them eventually, we'll see
  const stacks = Object.entries(groupedCoins);

  return (
    <div className="flex gap-2 pt-4 w-full max-w-full flex-wrap" data-testid="cash-hopper">
      {stacks.map(([denomination, stackSize]) => {
        return <div key={denomination} className="relative w-[3rem] h-[3rem]">
          {
            Array.from({ length: stackSize > 5 ? 5 : stackSize }, (_, index) => {
              return <Coin
                key={index}
                denomination={denomination}
                // generate for: top-[-3px] top-[-6px] top-[-9px] top-[-12px] top-[-15px]
                className={`absolute left-0 top-[${index * -3}px]`}
              />;
            })
          }
          {
            stackSize < 5
              ? (null)
              : (
                <div className="absolute bottom-[0.5rem] right-[-1px] rounded-full bg-white border border-black text-center w-4 h-4 text-xs">{stackSize}</div>
              )
          }
        </div>
      })}
    </div>
  )
}