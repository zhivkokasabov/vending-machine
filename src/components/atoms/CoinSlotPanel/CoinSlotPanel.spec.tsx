import { CoinSlotPanel } from ".";

import * as coinManager from "providers/CoinManager.provider";
import * as cashHopper from "providers/CashHopper.provider";
import { denominations } from "constants/denominations.const";
import { TooltipProvider } from "components/ui/tooltip";
import { fireEvent, render } from "@testing-library/react";

jest.mock("providers/CoinManager.provider");
jest.mock("providers/CashHopper.provider");

describe("CoinSlotPanel tests", () => {
  const insertCoinMock = jest.fn();
  const stashCoinsMock = jest.fn();

  const renderComponent = () => render(
   <TooltipProvider>
     <CoinSlotPanel />
   </TooltipProvider>
  );

  beforeEach(() => {
    jest.spyOn(coinManager, 'useCoinManager').mockReturnValue({
      insertCoin: insertCoinMock,
      reset: jest.fn(),
      total: 1
    });
    jest.spyOn(cashHopper, 'useCashHopper').mockReturnValue({
      stashCoins: stashCoinsMock,
      stashedCoins: []
    });
  });

  it("should render", () => {
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should call insertCoin with correct denomination", () => {
    const { getAllByRole } = renderComponent();

    const coinButtons = getAllByRole("button");

    fireEvent.click(coinButtons[2]);

    expect(insertCoinMock).toHaveBeenCalledTimes(1);
    expect(insertCoinMock).toHaveBeenCalledWith(denominations[2]);
  });

  it("should call stashCoins if invalid denomination", () => {
    const { getAllByRole } = renderComponent();

    const coinButtons = getAllByRole("button");

    fireEvent.click(coinButtons[8]);

    expect(stashCoinsMock).toHaveBeenCalledTimes(1);
  });
});
