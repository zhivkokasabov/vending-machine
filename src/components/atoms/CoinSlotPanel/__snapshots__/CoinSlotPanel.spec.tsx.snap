// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoinSlotPanel tests should render 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-3 gap-2"
  >
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
    >
      0.01
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
    >
      0.02
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
    >
      0.05
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
    >
      0.1
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
    >
      0.2
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
    >
      0.5
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
    >
      1
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 border border-neutral-200 bg-white shadow-sm hover:bg-neutral-100 hover:text-neutral-900 dark:border-neutral-800 dark:bg-neutral-950 dark:hover:bg-neutral-800 dark:hover:text-neutral-50 px-4 py-2 undefined rounded-full w-[3rem] h-[3rem]"
    >
      2
    </button>
    <button
      aria-label="Insert random coin"
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium transition-colors focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-neutral-950 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:focus-visible:ring-neutral-300 bg-red-500 text-neutral-50 shadow-sm hover:bg-red-500/90 dark:bg-red-900 dark:text-neutral-50 dark:hover:bg-red-900/90 px-4 py-2 rounded-full w-[3rem] h-[3rem]"
      data-state="closed"
    >
      ?
    </button>
  </div>
</DocumentFragment>
`;
